Userland/DevTools/UserspaceEmulator/SoftFPU.cpp

```diff
@@ -0,0 +1,994 @@
+/*
+ * Copyright (c) 2020, Andreas Kling <kling@serenityos.org>
+ * Copyright (c) 2021, Leon Albrecht <leon2002.la@gmail.com>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include "SoftFPU.h"
+#include "Emulator.h"
+#include "SoftCPU.h"
+#include "ValueWithShadow.h"
+
+#include <AK/BitCast.h>
+#include <unistd.h>
+
+#if defined(__GNUC__) && !defined(__clang__)
+#    pragma GCC optimize("O3")
+#endif
+
+#define TODO_INSN()                                                                   \
+    do {                                                                              \
+        reportln("\n=={}== Unimplemented instruction: {}\n", getpid(), __FUNCTION__); \
+        m_emulator.dump_backtrace();                                                  \
+        _exit(0);                                                                     \
+    } while (0)
+
+template<typename T>
+ALWAYS_INLINE void warn_if_uninitialized(T value_with_shadow, const char* message)
+{
+    if (value_with_shadow.is_uninitialized()) [[unlikely]] {
+        reportln("\033[31;1mWarning! Use of uninitialized value: {}\033[0m\n", message);
+        UserspaceEmulator::Emulator::the().dump_backtrace();
+    }
+}
+
+namespace UserspaceEmulator {
+
+long double SoftFPU::fpu_get(u8 index) const
+{
+    VERIFY(index < 8);
```

Comment:
Can the 8 be put into a `constexpr` variable instead of being a magic number?

Reply:
I have conflicting feelings about that, this limit is the size of the stack and should always be 8 and making it a variable, even a constexpr one will make it seem like it can be changed, which is not possible on a hardware level - there are only 3 bits reserved to index into this- so maybe not?